app/src/main/java/com/cmput301f20t21/bookfriends/ui/login/CreateAccountActivity.java

```diff
@@ -93,18 +93,85 @@ private boolean checkEmpty(TextInputLayout layout) {
+        String username = usernameField.getText().toString();
+        boolean notAlphanumeric = false;
+
+        if (!username.matches("[a-zA-Z0-9]+")) {
+            usernameLayout.setError(getString(R.string.alphanumeric_error));
+            notAlphanumeric = true;
+        }
+        return notAlphanumeric;
+    }
+
+    private boolean checkPasswordLength() {
+        EditText passwordField = passwordLayout.getEditText();
+        String password = passwordField.getText().toString();
+        boolean isIncorrectLength = false;
+
+        if (password.length() < 6) {
+            passwordLayout.setError(getString(R.string.password_length_error));
+            isIncorrectLength = true;
+        }
+
+        return isIncorrectLength;
+    }
+
+    private boolean isEmailValid() {
+        EditText emailField = emailLayout.getEditText();
+        String validEmail = emailField.getText().toString();
+        boolean valid = android.util.Patterns.EMAIL_ADDRESS.matcher(validEmail).matches();
+        if(!valid){
+            emailLayout.setError(getString(R.string.email_format_error));
+        }
+        return valid;
+    }
+
+    private boolean checkPasswordMatch() {
+        EditText passwordField = passwordLayout.getEditText();
+        String password = passwordField.getText().toString();
+        EditText confirmPasswordField = confirmLayout.getEditText();
+        String confirmPassword  = confirmPasswordField.getText().toString();
+        boolean isPasswordMatch = false;
+        if(!confirmPassword.equals(password)) {
+            confirmLayout.setError(getString(R.string.password_match_error));
+            isPasswordMatch = true;
+        }
+        return isPasswordMatch;
+    }
+
     public void onCancelClicked(View view) {
         finish();
     }
 
     public void onCreateClicked(View view){
-        String confirmPassword = confirmField.getText().toString();
         String password = passwordField.getText().toString();
-        if(!confirmPassword.equals(password)){
-            confirmLayout.setError(getString(R.string.password_match_error));
-        }
-        else if (!checkEmpty(usernameLayout) && !checkEmpty(passwordLayout) &&
-                !checkEmpty(confirmLayout) && !checkEmpty(emailLayout)) {
+        boolean checkEmailValid = isEmailValid();
+        boolean checkPasswordLength = checkPasswordLength();
+        boolean checkUsername = checkUsername();
+        boolean checkUsernameSpace = checkSpace(usernameLayout);
+        boolean checkPasswordSpace = checkSpace(passwordLayout);
+        boolean checkUsernameEmpty = checkEmpty(usernameLayout);
+        boolean checkConfirmEmpty = checkEmpty(confirmLayout);
+        boolean checkPasswordEmpty = checkEmpty(passwordLayout);
+        boolean checkEmailEmpty = checkEmpty(emailLayout);
+        boolean checkPasswordMatch = checkPasswordMatch();
+        if (checkEmailValid&&!checkPasswordLength&&!checkUsername&&!checkUsernameSpace
```

Comment:
The logic here looks messy to me because we have some checkers that returns true to be valid and some that's false to be valid (like `notAlphaNumeric` for checkUsername) 

suggestion: we need to inverse the logic and function name for them. i.e. checkUsername -> checkUsernameValid, and returns isAlphaNumeric. same for checkEmpty -> checkNotEmpty, etc. that way the if logic here will be very straight forward instead of seeing a lot of `!` logic

Reply:
Rather than name checkUsernameValid, I think isUsernameValid would be a better name. Check should be the function name and it returns isValid.